src/analysis.ts

```diff
@@ -36,10 +38,60 @@ export class ValidationError extends Error {
 
 
 export class Analysis {
-  descriptors: DocumentDescriptor[];
+  private _descriptors: DocumentDescriptor[];
+  private _elementsByTagName = new Map<string, ElementDescriptor>();
+  private _elementsByPackageDir = new Map<string, ElementDescriptor[]>();
+  private _behaviorsByIdentifierName = new Map<string, BehaviorDescriptor>();
+
+  elementPaths: Map<ElementDescriptor, string>;
+  behaviorPaths: Map<ElementDescriptor, string>;
 
   constructor(descriptors: DocumentDescriptor[]) {
-    this.descriptors = descriptors;
+    this._descriptors = descriptors;
+    const packageGatherer = new PackageGatherer();
+    const elementsGatherer = new ElementGatherer();
+    new AnalysisWalker(this._descriptors).walk([
+      packageGatherer, elementsGatherer
+    ]);
+
+    for (const element of elementsGatherer.elementDescriptors) {
+      if (element.is) {
+        this._elementsByTagName.set(element.is, element);
+      }
+      let elementPath = elementsGatherer.elementPaths.get(element);
+      const matchingPackageDirs =
+          Array.from(packageGatherer.packageDirs)
+              .filter(dir => elementPath.startsWith(dir));
+      const longestMatchingPackageDir =
+          matchingPackageDirs.sort((a, b) => b.length - a.length)[0] || '';
+
+      const elementsInPackage =
+          this._elementsByPackageDir.get(longestMatchingPackageDir) || [];
+      elementsInPackage.push(element);
+      this._elementsByPackageDir.set(
+          longestMatchingPackageDir, elementsInPackage);
+    }
+
+    for (const behavior of elementsGatherer.behaviorDescriptors) {
+      if (behavior.symbol) {
+        this._behaviorsByIdentifierName.set(behavior.symbol, behavior);
+      }
+    }
+
+    this.behaviorPaths = elementsGatherer.behaviorPaths;
+    this.elementPaths = elementsGatherer.elementPaths;
+  }
+
+  getElement(tag: string): ElementDescriptor|undefined {
+    return this._elementsByTagName.get(tag);
+  }
+
+  getElementsForPackage(dirName: string): ElementDescriptor[]|undefined {
+    return this._elementsByPackageDir.get(dirName);
+  }
+
+  getBehavior(identifier: string): BehaviorDescriptor|undefined {
```

Comment:
since behaviors can be referenced by name, or an expression, make sure you clarify what this method expects